common/reconciliation/invariant/staleWorkflow.go

```diff
@@ -0,0 +1,628 @@
+		last.WorkflowExecutionCompletedEventAttributes,
+		last.WorkflowExecutionContinuedAsNewEventAttributes,
+		last.WorkflowExecutionFailedEventAttributes,
+		last.WorkflowExecutionTerminatedEventAttributes,
+		last.WorkflowExecutionTimedOutEventAttributes,
+	) {
+		// completion event may simply not exist due to incomplete replication.
+		// this may be handled by history_exists or broken-history invariants, and we don't need to handle it here.
+		return when, false, c.failed("last event is not a completion-type", "missing data? got: %v", last)
+
+	}
+
+	if last.Timestamp == nil {
+		return when, false, c.failed("last event has a nil timestamp", "bad data? got: %v", last)
+	}
+	// sanity check, so far I have not seen any completion events in the top-level state.
+	// I suspect it's only an old / cache field, but bail if it disagrees, just in case.
+	if workflow.State.ExecutionInfo.CompletionEvent != nil &&
+		workflow.State.ExecutionInfo.CompletionEvent.Timestamp != nil &&
+		*last.Timestamp != *workflow.State.ExecutionInfo.CompletionEvent.Timestamp {
+		return when, false, c.failed(
+			"workflow execution info and final event disagree",
+			"bad data? got execution-completion-event timestamp: %v and final history event timestamp: %v",
+			time.Unix(0, *last.Timestamp), time.Unix(0, *workflow.State.ExecutionInfo.CompletionEvent.Timestamp))
+	}
+	ts := time.Unix(0, *last.Timestamp)
+	if ok, result := c.checkTimeInSaneRange(ts, "last event timestamp"); !ok {
+		return when, false, result
+	}
+	return ts, true, result
+}
+
+func anyPresent(items ...interface{}) bool {
+	for _, i := range items {
+		// cannot use `i == nil` because typed nil pointers are boxed into non-nil interface{} due to having a type
+		if !reflect.ValueOf(i).IsNil() {
+			fmt.Printf("not nil: %T\n", i)
+			return true
+		}
+	}
+	return false
+}
+
+func (c *staleWorkflowCheck) getLastEvent(branchToken []byte, domainName string) (*types.HistoryEvent, error) {
+	const (
+		maxHistoryLen = 250000 // our internal limits are much smaller, but it would be fine to raise this
+		pageSize      = 1000   // multiple 1000/page limits elsewhere, also fine to change
+		batchTimeout  = 5 * time.Second
+	)
+	iter := 0
+	var nextPageToken []byte
+	var lastEvent *types.HistoryEvent
+	for ; iter < (maxHistoryLen / pageSize); iter++ { // loose sanity check
+		ctx, cancel := context.WithTimeout(context.Background(), batchTimeout)
+		defer cancel() // revive:disable-line:defer clean up on panics, dups are just noops
+
+		shard := c.pr.GetShardID()
+		history, err := c.pr.ReadHistoryBranch(ctx, &persistence.ReadHistoryBranchRequest{
+			BranchToken: branchToken,
+			// only interested in the last event, but we have to read in order to get there.
+			MinEventID:    common.FirstEventID,
+			MaxEventID:    common.EndEventID,
+			ShardID:       &shard,
+			PageSize:      pageSize,
+			DomainName:    domainName,
+			NextPageToken: nextPageToken,
+		})
+		cancel() // handle non-panics
```

Comment:
you're cancelling twice?

Reply:
the deferred one would run at function exit, not on the next loop.

there are other ways to achieve this (mutating a var, deferring a late-bound call at the beginning), it just seemed simpler to call it twice.